source/non_matching/fe_values.cc

```diff
@@ -199,7 +218,36 @@ namespace NonMatching
 
         // Save the element and the local dof values, since this is what we need
         // to evaluate the function.
-        element = &dof_handler_cell->get_fe();
+
+        // Check if we can use the fast path. In case we have a different
+        // element from the one used before we need to set up the data
+        // structures again.
+        if (element != &dof_handler_cell->get_fe())
+          {
+            poly.clear();
+            element = &dof_handler_cell->get_fe();
+
+            if (element->n_base_elements() == 1 &&
+                dealii::internal::FEPointEvaluation::is_fast_path_supported(
```

Comment:
I was thinking about the same thing a couple of days ago, but would have opted towards teaching FEP to skip the application of the mapping. Are there other instances where we might want to skip the application of mapping?

Reply:
`FEPointEvaluation` is too heavy-weight here, as it involves several data structures and layers. Here, we need something really light-weight, as the use case is to be able to apply it inside various `Function` objects for a number of tasks, all applied with a `value`/`gradient`/`hessian` call for one point at a time (even the current code has overhead, but it is still around 10x faster for this particular task). My plan would rather be to factor out the evaluation part of `FEPointEvaluation`, i.e. the part that calls `evaluate_tensor_product_...`, the polynomials and the renumbering into a separate class and let both `FEPointEvaluation` and this one use it.

Reply:
Do you want me to change it here?

Reply:
>  Here, we need something really light-weight, as the use case is to be able to apply it inside various Function objects for a number of tasks, all applied with a value/gradient/hessian call for one point at a time (even the current code has overhead, but it is still around 10x faster for this particular task).

Yes, I did see it as well that the function is called for a single point. I was also investigating if it would make sense to overload the `Function` methods that take `std::vector<...>` instead and collect at the calling function multiple points. But could not figure out - due to lack of time - if that is even possible at the caller side.

> My plan would rather be to factor out the evaluation part of FEPointEvaluation, i.e. the part that calls evaluate_tensor_product_..., the polynomials and the renumbering into a separate class and let both FEPointEvaluation and this one use it.

That sounds perfect!

> Do you want me to change it here?

I don't mind if we do this change i a follow-up PR!